scalyr_agent/line_matcher.py

```diff
@@ -125,6 +169,135 @@ def _readline( self, file_like, max_length=0 ):
+
+        if current_file is None:
+            current_file = '<unknown>'
+
+        # read from the buffer
+        line = file_like.readline( max_length )
+
+        #return early if no line found
+        if len( line ) == 0:
+            return LogLine( '' ), False
+
+        # check for presence of newline chars to see if this is a partial line
+        partial = line[-1] != '\n' and line[-1] != '\r'
+
+        # early abort if this appears to be a partial line
+        if partial:
+            log.warn("Error parsing line as json for %s.  Log line does not appear to be a full line.  You can increase the buffer size for json line parsing with the global config option `read_page_size`.  The full line will be uploaded to Scalyr: %s" % (current_file, line.decode( "utf-8", 'replace' )),
+                     limit_once_per_x_secs=300, limit_key=('bad-json-%s' % current_file))
+
+            return self._truncated_log_line(line, self.max_line_length), partial
+
+        message = None
+        attrs = {}
+        timestamp = None
+
+        try:
+            json = scalyr_util.json_decode( line )
+
+            # go over all json key/values, adding non-message values to a attr dict
+            # and the message value to the message variable
+            for key, value in json.iteritems():
+                if key == self._json_message_field:
+                    message = value
+                elif key == self._json_timestamp_field:
+                    # TODO: need to add support for multiple timestamp formats
+                    timestamp = scalyr_util.rfc3339_to_nanoseconds_since_epoch(value)
+                    if 'raw_timestamp' not in attrs:
+                        attrs['raw_timestamp'] = value
+                else:
+                    attrs[key] = value
+
+        except Exception, e:
+            # something went wrong. Return the full line and log a message
+            log.warn("Error parsing line as json for %s.  Logging full line: %s\n%s" % (current_file, e, line.decode( "utf-8", 'replace' )),
+                     limit_once_per_x_secs=300, limit_key=('bad-json-%s' % current_file))
+            return self._truncated_log_line( line, self.max_line_length ), False
+
+
+        # if we didn't find a valid message key/value pair
+        # throw a warning and treat it as a normal line
+        if message is None:
+            log.warn("Key '%s' doesn't exist in json object for log %s.  Logging full line. Please check the log's 'json_message_field' configuration" %
+                (self._json_message_field, current_file), limit_once_per_x_secs=300, limit_key=('json-message-field-missing-%s' % current_file))
+            return self._truncated_log_line( line, self.max_line_length ), False
+
+        # we found a key match for the message field, so use that for the log message
+        # and store any other fields in the attr dict
+        result = self._truncated_log_line( message, self.max_line_length )
```

Comment:
So, if the line is longer than max_line, I believe we will just write out the first part and throw everything else away.  A better approach (and what we normally do I believe) is just to emit the entire line in chunks of max_line size.  So, we'd just return the first portion here, and remember the remaining part of the line for future calls.  Is that doable, or are we not keeping enough state around?